File: jinwoo/11주차/달리기 경주.py
# 20분?
# Lv 1
def solution(players, callings):
    # 순위를 저장하는 딕셔너리
    dic={}
    for i in range(len(players)):
        dic[players[i]] = i
        

    
    # callings
    for player in callings:
        # 불린 사람의 현재 순위
        rank = dic[player]
        # 앞 사람
        frontPlayer = players[rank-1]
        
        # 스왑
        players[rank], players[rank-1] = players[rank-1], players[rank]
        # 순위 변경
        dic[player], dic[frontPlayer] =  dic[frontPlayer], dic[player]

    return players


# callings의 사람이 현재 몇 등인지 파악해야 함
# callings를 등수(idx) 배열로 만들기
# players에서 kai를 O(n)이 아닌 시간에 찾는게 관건 -> 딕셔너리

File: jHyuk/week11/달리기 경주.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>

using namespace std;

void PrintRank(map<int, string> m) {
    for (int i = 0; i < m.size(); i++) {
        cout << m[i] << ' ';
    }
    cout << endl;
}

vector<string> solution(vector<string> players, vector<string> callings) {
    vector<string> answer;

    map<int, string> map1;      //순위와 이름
    map<string, int> map2;      //이름과 순위

    for (int i = 0; i < players.size(); i++) {
        map1[i] = players[i];
        map2[players[i]] = i;
    }

    for (int i = 0; i < callings.size(); i++) {
        int rank = map2[callings[i]];

        map2[callings[i]] -= 1;
        map2[map1[rank - 1]] += 1;

        string tmp = map1[rank];
        map1[rank] = map1[rank - 1];
        map1[rank - 1] = tmp;

        //PrintRank(map1);
    }

    for (int i = 0; i < map1.size(); i++) {
        answer.push_back(map1[i]);
    }

    return answer;
}